### PR TITLE
zig: 0.9.1 -> 0.10.1

### PR DESCRIPTION
**Pull request by strager**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
It does more than merely add Zig. Why not concentrate this on the llvm? After that I can rebase and test on my already open PR.

**Comment by strager** [comment]
> It does more than merely add Zig. Why not concentrate this on the llvm? After that I can rebase and test on my already open PR.

See my note at the top of the PR:
> **This pull request is based on https://github.com/NixOS/nixpkgs/pull/209536. Please only consider the last commit in my PR.**

**Review by andrewrk — APPROVED** [review]
Looks great. I like what you did with coreutils env.

**Comment by AndersonTorres** [comment]
> See my note at the top of the PR:

Sorry, I can't ignore the whole context.

If you are saying that I should pretend your PR contains the zig update and nothing else (therefore, you are not willing to take any responsibility about the other commits in this PR), then I should equally pretend this PR is superfluous and close it.

What does your "look only to the latest commit please" pull request bring to the table that can't be grafted on the already opened one?

**Comment by strager** [comment]
> If you are saying that I should pretend your PR contains the zig update and nothing else (therefore, you are not willing to take any responsibility about the other commits in this PR), [...]

This is not what I meant. I meant that only the last commit in this PR is really up for review. I also meant that this PR should merge only after #209536 merges. I should have worded it more clearly.

I wanted to express this in the GitHub PR tool, but it wouldn't let me pick #209536's branch as the base branch.

> [...], then I should equally pretend this PR is superfluous and close it.

Huh? I don't understand. Why should we pretend that my PR is superfluous? The purpose of this PR is to share the Zig update commit ([8054072](https://github.com/NixOS/nixpkgs/pull/210324/commits/805407266c9e7500c4b9b8d70779b4e3050ce282)) which I called out twice now.

You are free to ignore my PR, but I don't know why you're being what I perceive as hostile (e.g. suggesting my PR be closed).

> What does your "look only to the latest commit please" pull request bring to the table that can't be grafted on the already opened one?

The already-opened PR (#199947) was owned by you, so I could not modify it (I think). And even if I could, I think it'd be unprofessional for me to change your PR.

I tested my PR. Even better, CI can test my PR. If I exclude the LLVM commits (which is what I think you're suggesting), CI could not test my PR.

**Comment by andrewrk** [comment]
@AndersonTorres it is common for a PR to be opened in anticipation of another one being merged. This allows to move forward faster, since it can be reviewed ahead of time, and tested ahead of time, and then simply rebased once the prerequisite PR is merged.

You are not being asked to pretend anything; simply to wait until the other PR is merged before dealing with this PR. Or, if you feel generous, you can offer review on the last commit in this PR, potentially saving a future review round trip for yourself and for @strager.

**Comment by vegerot** [comment]
@sTRAGER have you considered setting the base branch of your PR to #209536 ? 

That has other weirdnesses too, like someone merging this PR before #209536 is merged would mess that PR up.  But it would accurately show the diff.  

Besides, if you believe this PR shouldn’t be merged _now_ (for some reason) then consider marking it as a Draft until then.

Just my opinion, but imo since this PR is getting so much pushback I would change the base branch and mark it as a Draft

**Comment by vegerot** [comment]
(or look into a tool like ReviewStack / Sapling) that is more capable of handling these kinds of reviews

**Comment by Et7f3** [comment]
> @AndersonTorres it is common for a PR to be opened in anticipation of another one being merged. This allows to move forward faster, since it can be reviewed ahead of time, tested ahead of time, and then rebased once the prerequisite PR is merged.

using commit of another PR also has the advantages to test two PR at once. We see that llvm if working fine (ci of my PR) + another project that use it the test suite of zig confirm llvm is working. Here I reviewed this PR and @sager will be able to work on the PR while waiting for llvm to be merged so multithreading :)

I tested to build and it seem to fail on x86_64-darwin I got this error from the log:
```
thread 47121201 panic: attempt to unwrap error: OSVersionDetectionFail
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/process.zig:373:46: 0x107b33362 in std.process.getEnvVarOwned (zig2)
        const result = os.getenv(key) orelse return error.EnvironmentVariableNotFound;
[ 90%] Building CXX object CMakeFiles/zigstage1.dir/src/stage1/heap.cpp.o
[ 91%] Building CXX object CMakeFiles/zigstage1.dir/src/stage1/ir.cpp.o
[ 91%] Building CXX object CMakeFiles/zigstage1.dir/src/stage1/ir_print.cpp.o
[ 92%] Building CXX object CMakeFiles/zigstage1.dir/src/stage1/mem.cpp.o
[ 92%] Building CXX object CMakeFiles/zigstage1.dir/src/stage1/os.cpp.o
                                             ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/process.zig:373:46: 0x107b33362 in std.process.getEnvVarOwned (zig2)
        const result = os.getenv(key) orelse return error.EnvironmentVariableNotFound;
                                             ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/process.zig:373:46: 0x107b33362 in std.process.getEnvVarOwned (zig2)
        const result = os.getenv(key) orelse return error.EnvironmentVariableNotFound;
                                             ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/process.zig:373:46: 0x107b33362 in std.process.getEnvVarOwned (zig2)
        const result = os.getenv(key) orelse return error.EnvironmentVariableNotFound;
                                             ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/os.zig:2812:19: 0x107aaac62 in std.os.mkdiratZ (zig2)
        .NOENT => return error.FileNotFound,
                  ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/os.zig:2767:9: 0x107aaa968 in std.os.mkdirat (zig2)
        return mkdiratZ(dir_fd, &sub_dir_path_c, mode);
        ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/fs.zig:1426:9: 0x107aaa876 in std.fs.Dir.makeDir (zig2)
:
        const result = os.getenv(key) orelse return error.EnvironmentVariableNotFound;
                                             ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/process.zig:373:46: 0x107b33362 in std.process.getEnvVarOwned (zig2)
        const result = os.getenv(key) orelse return error.EnvironmentVariableNotFound;
                                             ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/process.zig:373:46: 0x107b33362 in std.process.getEnvVarOwned (zig2)
        const result = os.getenv(key) orelse return error.EnvironmentVariableNotFound;
                                             ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/os.zig:2812:19: 0x107aaac62 in std.os.mkdiratZ (zig2)
        .NOENT => return error.FileNotFound,
                  ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/os.zig:2767:9: 0x107aaa968 in std.os.mkdirat (zig2)
        return mkdiratZ(dir_fd, &sub_dir_path_c, mode);
        ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/fs.zig:1426:9: 0x107aaa876 in std.fs.Dir.makeDir (zig2)
        try os.mkdirat(self.fd, sub_path, default_new_dir_mode);
        ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/os.zig:1745:22: 0x107c80c1d in std.os.openatZ (zig2)
            .PERM => return error.AccessDenied,
                     ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/fs.zig:1175:13: 0x107aa2b62 in std.fs.Dir.openFileZ (zig2)
            try os.openatZ(self.fd, sub_path, os_flags, 0);
            ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/fs.zig:1102:9: 0x107a9ead0 in std.fs.Dir.openFile (zig2)
        return self.openFileZ(&path_c, flags);
        ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/fs.zig:2041:20: 0x107cd04bb in std.fs.Dir.readFile (zig2)
        var file = try self.openFile(file_path, .{});
                   ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/zig/system/darwin/macos.zig:71:13: 0x107b2e147 in std.zig.system.darwin.macos.detect (zig2)
            return error.OSVersionDetectionFail;
            ^
/tmp/nix-build-zig-0.10.0.drv-0/source/lib/std/zig/system/NativeTargetInfo.zig:72:23: 0x107b2d527 in std.zig.system.NativeTargetInfo.detect (zig2)
            .macos => try darwin.macos.detect(&os),
                      ^
/private/tmp/nix-build-zig-0.10.0.drv-0/source/src/main.zig:4945:5: 0x107a78d82 in main.detectNativeTargetInfo (zig2)
    return std.zig.system.NativeTargetInfo.detect(cross_target);
    ^
/private/tmp/nix-build-zig-0.10.0.drv-0/source/src/main.zig:3902:29: 0x107a73c35 in main.cmdBuild (zig2)
        const target_info = try detectNativeTargetInfo(cross_target);
                            ^
/private/tmp/nix-build-zig-0.10.0.drv-0/source/src/main.zig:261:9: 0x1079ff26a in main.mainArgs (zig2)
        return cmdBuild(gpa, arena, cmd_args);
        ^
???:?:?: 0x1093f690c in ___zig_fail_unwrap (???)
/private/tmp/nix-build-zig-0.10.0.drv-0/source/src/stage1.zig:56:43: 0x1079fe268 in main (zig2)
        stage2.mainArgs(gpa, arena, args) catch unreachable;
                                          ^
???:?:?: 0x1149a451d in ??? (???)
???:?:?: 0x0 in ??? (???)
make[2]: *** [CMakeFiles/stage3.dir/build.make:71: CMakeFiles/stage3] Abort trap: 6
make[1]: *** [CMakeFiles/Makefile2:260: CMakeFiles/stage3.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I can't debug right now (because I already debug another Darwin failure) and it seems more zig specific. I have Darwin sandbox activated.

**Comment by AndersonTorres** [comment]
> This is not what I meant. I meant that only the last commit in this PR is really up for review. I also meant that this PR should merge only after #209536 merges. I should have worded it more clearly.

OK.

> Why should we pretend that my PR is superfluous?

In a certain sense, both PRs **cannot** be merged as is, because they modify the same files, have the same ending purpose, and any one of them can be accepted to the prejudice of the other. 

The only relevant difference between mine and yours is the cherry-picking of LLVM-related snippets that, according to you, are not the most relevant part of this PR.

At least in a large sense, one of them must go. Or, if you think this is too harsh, your PR supercedes or at least is equivalent to mine.

> I don't know why you're being what I perceive as hostile (e.g. suggesting my PR be closed).

You did not explain what was your purpose when you opened this PR. Your two messages 

https://github.com/NixOS/nixpkgs/pull/199947#issuecomment-1379951602

https://github.com/NixOS/nixpkgs/pull/210324#issue-1530279344

say absolutely nothing besides "hey I opened a PR".

>  And even if I could, I think it'd be unprofessional for me to change your PR.

You could cherry-pick the same way you did with the LLVM-related ones. At least I would be able compare the effective differences between them and hopefully merge a conjoined work.

File: pkgs/development/compilers/zig/default.nix
Comment (AndersonTorres) [inline pkgs/development/compilers/zig/default.nix:68]: Why are you changing from check to installCheck?

File: pkgs/development/compilers/zig/default.nix
Comment (AndersonTorres) [inline pkgs/development/compilers/zig/default.nix:41]:
```suggestion
  postPatch = ''
```

**Comment by strager** [comment]
@Et7f3 I tried on my AArch64 macOS machine. I get a different failure:
```
[snip]
warning(link): directory not found for '-L/nix/store/g8b81w9kpb82gm7702ca692mp5flm7
pf-apple-framework-CoreFoundation-11.0.0/lib'                                      
thread 24436935 panic: incorrect alignment                                         
/private/tmp/nix-build-zig-0.10.0.drv-0/source/src/link/MachO/Object.zig:114:72: 0x
101565863 in link.MachO.Object.parse (zig2)
                    @alignCast(@alignOf(macho.nlist_64), &self.contents[symtab.symo
ff]),
                                                                       ^
/private/tmp/nix-build-zig-0.10.0.drv-0/source/src/link/MachO/Archive.zig:230:21: 0
x10157a5af in link.MachO.Archive.parseObject (zig2)
    try object.parse(gpa, cpu_arch);
                    ^
[snip]
```
I'll debug what I see, and maybe that will fix the x86_64 build too.

**Comment by andrewrk** [comment]
0.10.1 will be tagged soon, within 1-2 weeks. This could potentially be fixed by one of the many [bug fixes that made it into that tag](https://github.com/ziglang/zig/milestone/18?closed=1).

**Comment by strager** [comment]
> You might consider testing with the 0.10.x branch.

I see the same problem with the 0.10.x branch, but the build works on my MacBook Pro on master (Zig commit 7cb2f9222da38d687e8708dd5d94d3175cc77995). I'm bisecting now.

File: pkgs/development/compilers/zig/default.nix
Comment (strager) [reply pkgs/development/compilers/zig/default.nix:68]: I don't know where `make` puts the stage3 build. It seems to be in zig-cache. But `make install` puts the stage3 build in a consistent place, so I decided to use installCheck.

**Comment by andrewrk** [comment]
Speaking for upstream, unfortunately, there are a lot of bug fixes in master branch that cannot be cherry-picked to 0.10.x because the branches have already diverged quite a bit. So unless the fix is a small patch that can be applied cleanly, we may be stuck with the bug in 0.10.1. I expect the quality of 0.11.0 to be much higher (this is related to [deleting the old compiler](https://ziglang.org/news/goodbye-cpp/) which happened *after* releasing 0.10.0).

**Comment by AndersonTorres** [comment]
If this minor release series will be "irredeemably buggy", should we bring the unstable/master release instead?

TBH I was planning this a long time ago, but there was no serious use for a master branch of zig.

**Comment by andrewrk** [comment]
My advice is to stick with packaging releases. They provide a point of coordination for applications and projects. For example, two relevant projects would be:

 * [ncdu](https://dev.yorhel.nl/ncdu) - uses zig 0.9.1
 * [river](https://github.com/riverwm/river) - uses zig 0.10.0

Applications which require a bleeding-edge compiler will likely prefer to obtain the compiler binary [directly from upstream](https://ziglang.org/download/).

**Comment by strager** [comment]
Bisect result: my Darwin bug is fixed between Zig commits [4e2a960b](https://github.com/ziglang/zig/commit/4e2a960b523070c7f8fddf0ea9b6e2a94e31dafe) (broken) and [20d86d9c](https://github.com/ziglang/zig/commit/20d86d9c63476b6312b87dc5b0e4aa4822eb7717) (working). Commits between fail because the stage1 compiler was ripped out. I don't know how to bisect further. I suspect further bisecting won't help, but maybe @andrewrk can figure out the issue.

```
# FAILS
rev = "4e2a960b523070c7f8fddf0ea9b6e2a94e31dafe";
hash = "sha256-1SpS0crLc1OuH+BP6EFWrfCWUapreAJU3fq8vO9HGBQ";

# WORKS
rev = "20d86d9c63476b6312b87dc5b0e4aa4822eb7717";
hash = "sha256-ZYdL9wPX1odppnkuOQHGacm5Z2JKfErbRSmc04WDF+Y";
```

**Comment by strager** [comment]
I have an idea: Maintain both 0.9.1 and 0.10.0:

* 0.9.1 is enabled for Darwin and Linux
* 0.10.0 is only enabled for Linux

If this sounds like a good idea, I can make the patch.

**Comment by strager** [comment]
I submitted a patch which keeps Zig 0.9.1 but defaults to Zig 0.10.0 on Linux. We needed Zig 0.9.1 anyway to avoid breaking packages such as ncdu.

File: pkgs/development/compilers/zig/0.10.0.nix
Comment (AndersonTorres) [inline pkgs/development/compilers/zig/0.10.0.nix:1]: If we are keeping the old release, then it is not an upgrade, but an addition.

So, split this commit into two:

1. the first one will rename the old `default.nix` file to a new `0.9.1.nix` (and updating `all-packages.nix` accordingly, to keep the "buildability");
2. the second one adding Zig 0.10, with a title:

```
zig_0_10: init at 0.10.0
```

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:23502]:
```suggestion
   zig_0_9 = darwin.apple_sdk_11_0.callPackage ../development/compilers/zig/0.9.nix { 
```

We are not keeping previous 0.9.x versions, so we don't need to include this in the expression name.

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:23506]:
```suggestion
  zig_0_10 = darwin.apple_sdk_11_0.callPackage ../development/compilers/zig/0.10.nix {
```

Same reason.

File: pkgs/top-level/all-packages.nix
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:29805]: Each of these changes should be in a separate commit, documenting why this is needed.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AndersonTorres** [comment]
@andrewrk, is there a policy of _Long Term Support_ on Zig? Or an update in a minor version is sufficient to consider it deprecated?

**Comment by andrewrk** [comment]
Zig follows [Semantic Versioning](https://semver.org/).

**Comment by edrex** [comment]
zig 0.10.1 released today https://github.com/ziglang/zig/releases/tag/0.10.1

Aside: I tried using this to package some Zig river utils (stacktile and levee) and they error out in the same way (`curl https://termbin.com/0ejca` with ANSI color) so I'm waiting on 0.10.1 to try again.

**Comment by Anton-4** [comment]
@AndersonTorres do you have the time to review this again or might someone else be able to help out?

We at [roc-lang](https://github.com/roc-lang) are also blocked on this PR and I would really appreciate your input :)

**Comment by AndersonTorres** [comment]
There are many blockers to solve here. The code does not even compile outside x64-linux.

**Comment by Anton-4** [comment]
I see, in any case, thanks for your quick reply!

@strager could it be that using zig 0.10.1 here would fix the issues on darwin?

**Comment by strager** [comment]
> @strager could it be that using zig 0.10.1 here would fix the issues on darwin?

Nope. Zig 0.10.1 on Darwin still has issues.

**Comment by strager** [comment]
> There are many blockers to solve here.

What blockers are you referring to?

**Comment by strager** [comment]
I updated this PR to use Zig 0.10.1 instead of Zig 0.10.0.

**Comment by wegank** [comment]
Please rebase :-)

**Comment by strager** [comment]
LLVM 15 landed in #194634. I rebased my PR on top of it.

**Comment by wegank** [comment]
@AndersonTorres Is this PR good to go?

**Comment by AndersonTorres** [comment]
LGTM

**Pull request merged**
